Unify quantized conv and linear tests

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26992 Unify quantized conv and linear tests**

Summary:
Run the same test for FBGEMM and QNNPACK backends.
Checks that QNNPACK or FBGEMM are supported before running it (using supported_qengines)

Test Plan:
    python test/test_quantized.py TestQuantizedLinear
    python test/test_quantized.py TestQuantizedConv
    python test/test_quantized_models.py
    python test/test_quantized_nn_mods.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17689171](https://our.internmc.facebook.com/intern/diff/D17689171)

supriyar committed:
Unify quantized conv and linear tests

Summary:
Run the same test for FBGEMM and QNNPACK backends.
Checks that QNNPACK or FBGEMM are supported before running it (using supported_qengines)

Test Plan:
python test/test_quantized.py TestQuantizedLinear
----------------------------------------------------------------------
Ran 2 tests in 0.108s

OK

python test/test_quantized.py TestQuantizedConv
----------------------------------------------------------------------
Ran 2 tests in 0.108s

OK

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify quantized conv and linear tests"


Summary:
Run the same test for FBGEMM and QNNPACK backends.
Checks that QNNPACK or FBGEMM are supported before running it (using supported_qengines)

Test Plan:
python test/test_quantized.py TestQuantizedLinear
----------------------------------------------------------------------
Ran 2 tests in 0.108s

OK

python test/test_quantized.py TestQuantizedConv
----------------------------------------------------------------------
Ran 2 tests in 0.108s

OK

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify quantized conv and linear tests"


Summary:
Run the same test for FBGEMM and QNNPACK backends.
Checks that QNNPACK or FBGEMM are supported before running it (using supported_qengines)

Test Plan:
python test/test_quantized.py TestQuantizedLinear
----------------------------------------------------------------------
Ran 2 tests in 0.108s

OK

python test/test_quantized.py TestQuantizedConv
----------------------------------------------------------------------
Ran 2 tests in 0.108s

OK

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify quantized conv and linear tests"


Summary:
Run the same test for FBGEMM and QNNPACK backends.
Checks that QNNPACK or FBGEMM are supported before running it (using supported_qengines)

Test Plan:
python test/test_quantized.py TestQuantizedLinear
----------------------------------------------------------------------
Ran 2 tests in 0.108s

OK

python test/test_quantized.py TestQuantizedConv
----------------------------------------------------------------------
Ran 2 tests in 0.108s

OK

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify quantized conv and linear tests"


Summary:
Run the same test for FBGEMM and QNNPACK backends.
Checks that QNNPACK or FBGEMM are supported before running it (using supported_qengines)

Test Plan:
python test/test_quantized.py TestQuantizedLinear
----------------------------------------------------------------------
Ran 2 tests in 0.108s

OK

python test/test_quantized.py TestQuantizedConv
----------------------------------------------------------------------
Ran 2 tests in 0.108s

OK

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dzhulgakov commented on test/test_quantized.py:
```diff
@@ -1089,420 +1089,422 @@ def test_qlinear(self, batch_size, input_channels, output_channels,
+    # Assert equal
+    if qengine == 'qnnpack':
+        np.testing.assert_array_almost_equal(
+            Y_q_ref2.int_repr().numpy(), Y_q.int_repr().numpy(), decimal=0)
```
maybe just set `decimal` in the beginning to 0 or None (or whatever default value) and use it everywhere?

dzhulgakov commented on test/test_quantized.py:
```diff
@@ -1089,420 +1089,422 @@ def test_qlinear(self, batch_size, input_channels, output_channels,
-            np.testing.assert_equal(W_q.q_zero_point(), W_unpacked.q_zero_point())
+    def test_qconv(self):
+        if 'qnnpack' in torch.backends.quantized.supported_engines:
+            if not IS_WINDOWS and not IS_PPC and not TEST_WITH_UBSAN:
```
why not windows? :) if qnnpack is broken - we should just not build it on windows

dzhulgakov commented on test/test_quantized.py:
```diff
@@ -1089,420 +1089,422 @@ def test_qlinear(self, batch_size, input_channels, output_channels,
+    def test_qconv(self):
+        if 'qnnpack' in torch.backends.quantized.supported_engines:
+            if not IS_WINDOWS and not IS_PPC and not TEST_WITH_UBSAN:
+                with enable_mobile_quantized_engine():
```
I'd change it to `with override_quantized_engine("qnnpack")` and ensure that fbgemm tests are surrounded by it too.

If there are too many tests to surround - you can also look into parametrized tests or just inheritance

dzhulgakov commented on test/common_utils.py:
```diff
@@ -185,7 +185,7 @@ def _check_module_exists(name):
 TEST_WITH_UBSAN = os.getenv('PYTORCH_TEST_WITH_UBSAN', '0') == '1'
 TEST_WITH_ROCM = os.getenv('PYTORCH_TEST_WITH_ROCM', '0') == '1'
-TEST_WITH_QNNPACK = os.getenv('PYTORCH_TEST_WITH_QNNPACK', '0') == '1'
+TEST_WITH_QNNPACK = 'qnnpack' in torch.backends.quantized.supported_engines
```
let's remove it from here and make it an inline condition

supriyar committed:
Update on "Unify quantized conv and linear tests"


Summary:
Run the same test for FBGEMM and QNNPACK backends.
Checks that QNNPACK or FBGEMM are supported before running it (using supported_qengines)

Test Plan:
python test/test_quantized.py TestQuantizedLinear
----------------------------------------------------------------------
Ran 2 tests in 0.108s

OK

python test/test_quantized.py TestQuantizedConv
----------------------------------------------------------------------
Ran 2 tests in 0.108s

OK

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dskhudia commented on test/test_quantized.py:
```diff
@@ -1089,423 +1088,425 @@ def test_qlinear(self, batch_size, input_channels, output_channels,
-           use_bias=st.booleans(),
-           use_relu=st.booleans(),
-           use_multi_dim_input=st.booleans(),
-           use_channelwise=st.booleans())
```
Why not just add another hypothesis parameter here qengine and select "fbgemm" and "qnnpack" for that parameter?

supriyar replied:
Either way works. I took this approach as we need to check other conditions for qnnpack like `if not IS_PPC and not TEST_WITH_UBSAN:` and skip calling the test if either of these conditions are set.

dskhudia replied:
We can add those check inside the test itself and make it pass. IMO, without the context manager, it would be easier to read.

supriyar replied:
We would still need the context manager to set the correct backend. But I guess it can be moved inside the test.

supriyar committed:
Update on "Unify quantized conv and linear tests"


Summary:
Run the same test for FBGEMM and QNNPACK backends.
Checks that QNNPACK or FBGEMM are supported before running it (using supported_qengines)

Test Plan:
    python test/test_quantized.py TestQuantizedLinear
    python test/test_quantized.py TestQuantizedConv
    python test/test_quantized_models.py
    python test/test_quantized_nn_mods.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify quantized conv and linear tests"


Summary:
Run the same test for FBGEMM and QNNPACK backends.
Checks that QNNPACK or FBGEMM are supported before running it (using supported_qengines)

Test Plan:
    python test/test_quantized.py TestQuantizedLinear
    python test/test_quantized.py TestQuantizedConv
    python test/test_quantized_models.py
    python test/test_quantized_nn_mods.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify quantized conv and linear tests"


Summary:
Run the same test for FBGEMM and QNNPACK backends.
Checks that QNNPACK or FBGEMM are supported before running it (using supported_qengines)

Test Plan:
    python test/test_quantized.py TestQuantizedLinear
    python test/test_quantized.py TestQuantizedConv
    python test/test_quantized_models.py
    python test/test_quantized_nn_mods.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17689171](https://our.internmc.facebook.com/intern/diff/D17689171)

[ghstack-poisoned]

jerryzh168 commented on test/test_quantized_models.py:
```diff
@@ -1,33 +1,37 @@
+                    continue
+            with override_quantized_engine(qengine):
+                torch.manual_seed(42)
+                myModel = ModelMultipleOps().to(torch.float32)
```
nit: can you follow the python naming convention? i.e. changing this to `my_model`

jerryzh168 commented on test/test_quantized_models.py:
```diff
@@ -1,33 +1,37 @@
-        # Setting target SQNR to be 30 dB so that relative error is 1e-3 below the desired
-        # output
-        self.assertGreater(SQNRdB, 30, msg='Quantized model numerics diverge from float, expect SQNR > 30 dB')
+        for qengine in ["fbgemm", "qnnpack"]:
```
Also, this is an integration test, I think you should put this in torch.quantization, but this can be done in a separate PR

supriyar replied:
cc @raghuramank100

jerryzh168 commented on test/test_quantized_models.py:
```diff
@@ -1,33 +1,37 @@
-        # output
-        self.assertGreater(SQNRdB, 30, msg='Quantized model numerics diverge from float, expect SQNR > 30 dB')
+        for qengine in ["fbgemm", "qnnpack"]:
+            if qengine not in torch.backends.quantized.supported_engines:
```
Also, shouldn't we use randomized tests to test numerics?

supriyar replied:
I haven't modified the test really. Just run the existing test for different qengine. 
cc @raghuramank100 for adding randomized tests.

jerryzh168 commented on test/test_quantized_models.py:
```diff
@@ -51,60 +55,74 @@ def test_float_quant_compare_per_channel(self):
+            if qengine not in torch.backends.quantized.supported_engines:
+                continue
+            if qengine == 'qnnpack':
+                if IS_PPC or TEST_WITH_UBSAN:
```
nit: this can be merged with the previous if

jerryzh168 reviewed (approved):
LG, some comments are inline

supriyar committed:
Update on "Unify quantized conv and linear tests"


Summary:
Run the same test for FBGEMM and QNNPACK backends.
Checks that QNNPACK or FBGEMM are supported before running it (using supported_qengines)

Test Plan:
    python test/test_quantized.py TestQuantizedLinear
    python test/test_quantized.py TestQuantizedConv
    python test/test_quantized_models.py
    python test/test_quantized_nn_mods.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17689171](https://our.internmc.facebook.com/intern/diff/D17689171)

[ghstack-poisoned]
